### ISSUE TITLE
Modify clang-format to support Bill Joy Normal Form

### ISSUE THREAD

**Issue by ryao** [openzfs/zfs#15791]
### Describe the feature would like to see added to OpenZFS

This is not actually an addition to OpenZFS itself, but to the external LLVM project. The idea is that we should be able to do something like `clang-format -i -style=bjnf path/to/file.c` and then have all cstyle.pl complaints corrected automatically.

### How will this feature improve OpenZFS?

Not all of us always write code perfectly in Bill Joy Normal Form. After writing a big change, we sometimes have complaints from cstyle.pl. It would be nice if we could just run a tool to fix them for us instead of fixing it by hand.

Since this is most relevant to OpenZFS, I am filing the idea here to share with others in the hope that a volunteer picks it up.

### Additional context

https://en.wikipedia.org/wiki/Kernel_Normal_Form

**Comment by ryao** [comment]
The clang-format tool is configurable:

https://clang.llvm.org/docs/ClangFormatStyleOptions.html

It seems that we could just define a style in .clang-format and then use `clang-format -i -style=file path/to/file.c`. The tricky part is defining the style. We would want a .clang-format file that leaves all of the files in bjnf untouched when clang-format is run on them using it. The imported lua code and zstd code does not follow bjnf, so we would not want to use those to evaluate the correctness of the style.

**Comment by ryao** [comment]
I am told in IRC that someone already started this:

https://github.com/vladak/clang-format-SunOS/blob/master/.clang-format

**Comment by rincebrain** [comment]
I don't think you can actually express everything cstyle is upset about in `clang-format` right now.

In particular:
```(void) foo;
bar((uint64_t)baz);
sizeof (quot);
```
I do not think you can express all of those at once right now. Expressing `sizeof (foo)` has side effects of making you put those spaces in _every_ statement (so we'd end up with `baz ((uint64_t)foo);`), and the casting directives don't have a special case for `(void) foo;` AFAICT.

So I'm probably going to end up trying to get everyone to agree on something `clang-format` can express reliably, since I just spent longer on a patch formatting it for this than I did on writing and testing it, and even if we convince LLVM upstream to add support for all our quirks, we'd then be stuck not being able to use `clang-format` until it made it into stable releases.

**Comment by robn** [comment]
Yeah, that's the right approach. The actual rules don't really matter (within reason) as much as them always being the same and easily repeatable. And then expressing 9x% of the current rules and saying "dang, tools don't support the rest" is kind of a winning argument for the remainder (not that I can imagine too many people dying on a hill of one extra space). And then we reformat, one super ugly "reformat the world" commit, and that's that. We won't be the first project  to have gone through it.

Big :+1: from me.
